lsp-mode.el

```diff
@@ -79,6 +79,45 @@
                 "lsp-define-client: :ignore-regexps element %s is not a string"
                 e))))
 
+;; --------------------------------------------------------
+
+(cl-defmacro lsp-define-whitelist-enable (name get-root
+                                               &key docstring)
+  "Define a function to add the project root for the current buffer to the whitleist.
+NAME is the base name for the command.
+GET-ROOT is the language-specific function to determint the project root for the current buffer."
+  (let ((enable (intern (format "%s-whitelist-add" name))))
+    `(defun ,enable ()
+       ,docstring
+       (interactive)
+       (let ((root (funcall ,get-root)))
+         (customize-save-variable 'lsp-project-whitelist
+                                    (add-to-list 'lsp-project-whitelist root))))))
+
+(cl-defmacro lsp-define-whitelist-disable (name get-root
+                                               &key docstring)
+  "Define a function to remove the project root for the current buffer from the whitleist.
+NAME is the base name for the command.
+GET-ROOT is the language-specific function to determint the project root for the current buffer."
+  (let ((enable (intern (format "%s-whitelist-remove" name))))
+    `(defun ,enable ()
+       ,docstring
+       (interactive)
+       (let ((root (funcall ,get-root)))
+         (customize-save-variable 'lsp-project-whitelist
+                                  (remove root 'lsp-project-whitelist ))))))
```

Comment:
Here I would stop the client if the project was removed.

Reply:
I don' think there is an `lsp-MAJOR-MODE-disable` at the moment.

Reply:
You could call `lsp--shutdown-cur-workspace`.

Reply:
I just tried that, and got a mountain of problems. I suspect that the shutdown is not all that solid, tbh.

I would rather put this in without that feature, and it can be added later as part of sorting out the shutdown.

This thing is complex enough already.

Reply:
Yeah :) You are right.